definition_validator/variance.cc

```diff
@@ -0,0 +1,184 @@
+                    // type of a method alias.
+                    ENFORCE(variance == core::Variance::CoVariant);
+
+                    // TODO: it's not clear what should be done here, as any
+                    // type_member references would be invalid from this
+                    // context.
+                    return;
+                } else {
+                    Exception::raise("Unhandled type alias: {}", alias->toString(ctx));
+                }
+            },
+
+            [&](core::ShapeType *shape) {
+                for (auto value : shape->values) {
+                    validate(ctx, variance, value);
+                }
+            },
+
+            [&](core::TupleType *tuple) {
+                for (auto value : tuple->elems) {
+                    validate(ctx, variance, value);
+                }
+            },
+
+            [&](core::AppliedType *app) {
+                auto members = app->klass.data(ctx)->typeMembers();
+                auto params = app->targs;
+
+                // We expect that types are fully saturated.
+                ENFORCE(members.size() == params.size());
+
+                for (int i = 0; i < members.size(); ++i) {
+                    auto memberVariance = members[i].data(ctx)->variance();
+                    auto typeArg = params[i];
+
+                    // Invert the variance when the type_member was defined as
+                    // contravariant (:in).
+                    const core::Variance checkedVariance =
+                        (variance == core::Variance::ContraVariant) ? invertVariance(memberVariance) : memberVariance;
+
+                    validate(ctx, checkedVariance, typeArg);
+                }
+            },
+
+            // This is where the actual variance checks are done.
+            [&](core::LambdaParam *param) {
+                auto paramData = param->definition.data(ctx);
+
+                ENFORCE(paramData->isTypeMember());
+
+                auto paramVariance = paramData->variance();
+
+                if (!compatibleVariance(variance, paramVariance)) {
+                    if (auto e = ctx.state.beginError(this->loc, core::errors::Resolver::InvalidVariance)) {
+                        auto flavor =
+                            paramData->owner.data(ctx)->isSingletonClass(ctx) ? "type_template" : "type_member";
+
+                        auto paramName = paramData->name.data(ctx)->show(ctx);
+
+                        e.setHeader("`{}` `{}` was defined as `{}` but is used in an `{}` context", flavor, paramName,
+                                    showVariance(paramVariance), showVariance(variance));
+
+                        e.addErrorLine(paramData->loc(), "`{}` `{}` defined here as `{}`", flavor, paramName,
+                                       showVariance(paramVariance));
+                    }
+                }
+            },
+
+            [&](core::MetaType *mt) { validate(ctx, variance, mt->wrapped); },
```

Comment:
is this ever encountered? 
Let's convert this to `raise("should never happen")`